docs/rfcs/034-storage-controller-restarts.md

```diff
@@ -0,0 +1,207 @@
+* helm charts
+
+## Terminology
+
+* Observed State: in-memory mapping between tenant shards and their current pageserver locations - currently built up
+at start-up by quering pageservers
+* Deployment: Kubernetes [primitive](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) that models
+a set of replicas
+
+## Implementation
+
+### High Level Flow
+
+At a very high level the proposed idea is to start a new storage controller instance while
+the previous one is still running and cut-over to it when it becomes ready. The new instance,
+should coordinate with the existing one and transition responsibility gracefully. While the controller
+has built in safety against split-brain situations (via generation numbers), we'd like to avoid such
+scenarios since they can lead to availability issues for tenants that underwent changes while two controllers
+were operating at the same time and require operator intervention to remedy.
+
+### Kubernetes Deployment Configuration
+
+On the Kubernetes configuration side, the proposal is to update the storage controller `Deployment`
+to use `spec.strategy.type = RollingUpdate`, `spec.strategy.rollingUpdate.maxSurge=1` and `spec.strategy.maxUnavailable=0`.
+Under the hood, Kubernetes creates a new replica set and adds one pod to it (`maxSurge=1`). The old replica set does not
+scale down until the new replica set has one replica in the ready state (`maxUnavailable=0`).
+
+The various possible failure scenarios are investigated in the [Handling Failures](#handling-failures) section.
+
+### Storage Controller Start-Up
+
+This section describes the primitives required on the storage controller side and the flow of the happy path.
+
+#### Database Table For Leader Synchronization
+
+A new table should be added to the storage controller database for leader synchronization during startup.
+This table will always contain at most one row. The proposed name for the table is `leader` and the schema
+contains two elements:
+* `hostname`: represents the hostname for the current storage controller leader - should be addressible
+from other pods in the deployment
+* `start_timestamp`: holds the start timestamp for the current storage controller leader (UTC timezone) - only required
+for failure case handling: see [Previous Leader Crashes Before New Leader Readiness](#previous-leader-crashes-before-new-leader-readiness)
+
+Storage controllers will read the leader row at start-up and then update it to mark themselves as the leader
+at the end of the start-up sequence. We want compare-and-exchange semantics for the update: avoid the
+situation where two concurrent updates succeed and overwrite each other. The default Postgres isolation
+level is `READ COMMITTED`, which isn't strict enough here. This update transaction should use the `REPEATABLE
+READ` isolation level in order to [prevent lost updates](https://www.interdb.jp/pg/pgsql05/08.html):
+
+```
+START TRANSACTION ISOLATION LEVEL REPEATABLE READ
```

Comment:
I feel the transaction has to start at the initial reading of the leaders table? We might need to do:

- Read leader if exists
```sql
START TRANSACTION ISOLATION LEVEL REPEATABLE READ
SELECT hostname, start_timestamp FROM leader LIMIT 1;
```
- ask leader to `/step_down` via HTTP.

- Update table through compare and exchange.

```sql
UPDATE leader SET hostname=<new_hostname>, start_timestamp=<new_start_ts>
WHERE hostname=<old_hostname>, start_timestamp=<old_start_ts>;
```

Reply:
Using `FOR UPDATE` clause under `READ-COMMITTED` is also an alternative than changing it to `REPEATABLE READ`. e.g. `SELECT hostname, start_timestamp FROM leader FOR UPDATE;`

Personally I don't have a preference. 
https://www.postgresql.org/docs/9.3/sql-select.html#SQL-FOR-UPDATE-SHARE

Reply:
> When it tries to mark itself as the leader, the update will not match any rows and the pod will exit.

Yep getting a stale result is ok in our use case. 

>  A very slight advantage is that the isolation level is more generic (portable) than the row level locks.

Agree on this point. Let's use repeatable read.